### PR TITLE
Support TypeScript

### LINKED ISSUE #51: Revert to ES5
Hi base-x team! It is a simple error so I wont make it lengthy. I met some kind of this error ```Failed to minify the code from this file``` and it was base-x changed to use es6, to fix this error, could the team support babel or at least, just code it in es5 style?
Thanks team so much.

### PR DESCRIPTION
**Pull request by junderw**
1. Commit JS source for verifiability.
2. Target ES5 closes #51 

Javascript transpilers will some day insert private key stealing virus into some React Native app and people will wonder what happened......

But for now, as long as we commit and track the JS changes, we can verify that nothing is wrong on OUR end.

Now types are supported too, hooray.

### CODE REVIEW

**Comment by junderw** [comment]
I found that copy pasting the old index.js and the new transpiled index.js into a diff checking GUI app like meld was helpful... it highlighted the changes which is mostly just const and let removal.

**Comment by junderw** [comment]
will tag this with v3.0.6 and publish once I can get an ACK from someone.

File: package.json [hunk 1]
```diff
@@ -21,21 +21,27 @@
   "license": "MIT",
   "author": "Daniel Cousens",
   "files": [
-    "index.js"
+    "src"
```
Comment (sublimator) [inline package.json]: `ts_src` -> `src` and put compile output to `dist/` ?

File: ts_src/index.ts [hunk 2]
```diff
@@ -4,9 +4,10 @@
 // Distributed under the MIT software license, see the accompanying
 // file LICENSE or http://www.opensource.org/licenses/mit-license.php.
 
-const Buffer = require('safe-buffer').Buffer
+// @ts-ignore
+const _Buffer = require('safe-buffer').Buffer;
```
Comment (sublimator) [inline ts_src/index.ts]: You can probably get the type checking to work somehow via `@types/node`

File: package.json (hunk 1 above)
Comment (sublimator) [reply package.json]: this way keeps the git history better ?

File: package.json [hunk 3]
```diff
@@ -21,21 +21,27 @@
...
   ],
-  "main": "index.js",
+  "main": "src/index.js",
+  "types": "src/index.d.ts",
   "repository": {
     "type": "git",
     "url": "https://github.com/cryptocoinjs/base-x.git"
   },
   "scripts": {
+    "build": "tsc -p ./tsconfig.json ; standard --fix",
```
Comment (sublimator) [inline package.json]: Why `;` and not `&&` ? 
Which folder is standard scoped ? src ? ts_src ?

File: package.json (hunk 1 above)
Comment (junderw) [reply package.json]: No.

File: package.json (hunk 3 above)
Comment (junderw) [reply package.json]: standard checks all *.js files recursively in the folder.

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    "strictBindCallApply": true,
+    "strictPropertyInitialization": true,
+    "noImplicitThis": true,
+    "alwaysStrict": true,
+    "esModuleInterop": false,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true
+  },
+  "include": [
+      "ts_src/**/*.ts"
```
Comment (sublimator) [inline tsconfig.json]: tip: simply `ts_src` actually seems to work I discovered recently
whether you want to be explicit or not is another story

File: package.json (hunk 3 above)
Comment (junderw) [reply package.json]: Also, ; because even when tsc build fails it will still actually build the JS... so we should fix it.

Comment (sublimator) [reply package.json]: Hrmmm, don't you want it (CI etc, not sure what the setup here is) to fail if typechecking fails ?

**Comment by junderw** [comment]
Published as v3.0.6

**Pull request merged**